blocks/search.py

```diff
@@ -0,0 +1,266 @@
+            is reached.
+
+        Returns
+        -------
+        Most probable sequences, corresponding probabilities and costs.
+
+        """
+        if not self.compiled:
+            self.compile()
+        # Inputs repeated beam_size times
+        aux_inputs = OrderedDict(
+            [(name, numpy.tile(val, (1, self.beam_size)))
+             for name, val in inputs_val_dict.iteritems()])
+
+        # Precompute contexts
+        contexts = self.compute_contexts(aux_inputs)
+
+        cur_states = self.compute_initial_states(contexts)
+
+        cur_states['cur_outputs'] = cur_states['outputs']
+        cur_states['cur_outputs_mask'] = numpy.ones_like(
+            cur_states['cur_outputs'])
+        cur_states['cur_probs'] = numpy.ones_like(
+            cur_states['cur_outputs'])
+
+        for i in range(max_length):
+            cur_states.update(self.compute_next(contexts, cur_states))
+            next_probs = (cur_states['cur_probs'][:, :, None] *
+                          cur_states['probs'] *
+                          cur_states['cur_outputs_mask'][-1, :, None])
+
+            # Top probs
+            indexes, top_probs = self._top_probs(next_probs[:, :],
+                                                 self.beam_size,
+                                                 unique=i == 0)
+            cur_states['cur_probs'] = numpy.array(top_probs).T[None, :]
+            indexes = numpy.array(indexes)  # chunk, 2, beam
+            # current_outputs.
+            # here we suppose, that we have 2d outputs
+            outputs = indexes[1, :].copy()
+
+            # rearrange outputs
+            rearrange_ind = indexes[0, :]
+            for name in cur_states:
+                cur_states[name] = self._rearrange(cur_states[name],
+                                                   rearrange_ind)
+            for name in contexts:
+                contexts[name] = self._rearrange(contexts[name],
+                                                 rearrange_ind)
+
+            # construct next output
+            outputs = outputs[None, :]
+            cur_states['cur_outputs'] = numpy.append(cur_states['cur_outputs'],
+                                                     outputs.copy(), axis=0)
+            # check if we meet eol
+            next_out_mask = numpy.ones((1, self.beam_size),
+                                       dtype=floatX)
+
+            next_out_mask[0, :] = ((outputs[0, :] != eol_symbol) *
+                                   cur_states['cur_outputs_mask'][-1, :])
+            cur_states['cur_outputs_mask'] = numpy.append(
+                cur_states['cur_outputs_mask'],
+                next_out_mask.copy(),
+                axis=0)
+
+            # All first element in sequences ended
+            if numpy.all(cur_states['cur_outputs'][-1, 0] == eol_symbol):
+                break
+
+        # Select only best
+        best_outputs = cur_states['cur_outputs'][1:, 0]
```

Comment:
I am not sure what should be the desired output format for this method. I would say that a list of the candidates sequences without eol symbols and a list of respective costs is the best. That seems more convenient to work with compared to a dense matrix and a mask, apart from the case when we want to feed this data as an input batch. 

@pbrakel , what's your opinion?